lib/shape.ml

```diff
@@ -29,8 +32,36 @@ let rectangle ?x ?y length width =
 
 let ellipse ?x ?y rx ry =
   match (x, y) with
-  | Some x, Some y -> Ellipse {c = {x; y}; rx; ry}
-  | _ -> Ellipse {c = { x = canvas_mid.x; y = canvas_mid.y}; rx; ry}
+  | Some x, Some y -> Ellipse { c = { x; y }; rx; ry }
+  | _ -> Ellipse { c = { x = canvas_mid.x; y = canvas_mid.y }; rx; ry }
+
+let triangle ?x1 ?y1 ?x2 ?y2 ?x3 ?y3 () =
+  let p1 =
+    match (x1, y1) with
+    | Some x, Some y -> { x; y }
+    | _ -> canvas_mid
+  in
+  let p2 =
+    match (x2, y2) with
+    | Some x, Some y -> { x; y }
+    | _ -> canvas_mid
+  in
+  let p3 =
+    match (x3, y3) with
+    | Some x, Some y -> { x; y }
+    | _ -> canvas_mid
+  in
+  Triangle { p1; p2; p3 }
+
+let translate dx dy shape =
+  match shape with
+  | Circle circle -> Circle { circle with c = { x = circle.c.x + dx; y = circle.c.y + dy } }
+  | Rectangle rectangle -> Rectangle { rectangle with c = { x = rectangle.c.x + dx; y = rectangle.c.y + dy } }
+  | Ellipse ellipse -> Ellipse { ellipse with c = { x = ellipse.c.x + dx; y = ellipse.c.y + dy } }
+  | Triangle triangle ->
```

Comment:
To reiterate: please remove all the triangle details from this PR. Let's not mix two different features in a single PR. :)